crates/pallet-domains/src/lib.rs

```diff
@@ -368,6 +425,62 @@ mod pallet {
             Ok(())
         }
 
+        #[pallet::call_index(5)]
+        #[pallet::weight(Weight::from_all(10_000))]
+        // TODO: proper benchmark
+        pub fn submit_bundle_v2(
+            origin: OriginFor<T>,
+            opaque_bundle: OpaqueBundleOf<T>,
+        ) -> DispatchResult {
+            ensure_none(origin)?;
+
+            log::trace!(target: "runtime::domains", "Processing bundle: {opaque_bundle:?}");
+
+            let domain_id = opaque_bundle.domain_id();
+            let bundle_hash = opaque_bundle.hash();
+            let extrinsics_root = opaque_bundle.extrinsics_root();
+            let operator_id = opaque_bundle.operator_id();
+            let (bundle_author, receipt) = opaque_bundle.deconstruct();
+
+            let receipt_type = execution_receipt_type::<T>(domain_id, &receipt);
+            if receipt_type == ReceiptType::Stale {
+                // The stale receipt should not be further processed, but we still track them for purposes
+                // of measuring the bundle production rate.
+                Self::note_domain_bundle(domain_id);
+                return Ok(());
+            }
+
+            // Add the exeuctione receipt to the block tree
+            process_execution_receipt::<T>(domain_id, operator_id, receipt, receipt_type)
+                .map_err(Error::<T>::from)?;
+
+            // `SuccessfulBundles` is empty means this is the first accepted bundle for this domain in this
+            // consensus block, which also mean a domain block will be produced thus update `HeadDomainNumber`
+            // to this domain block's block number.
+            if SuccessfulBundles::<T>::get(domain_id).is_empty() {
+                let next_number = HeadDomainNumber::<T>::get(domain_id)
+                    .checked_add(&One::one())
+                    .ok_or::<Error<T>>(BlockTreeError::MaxHeadDomainNumber.into())?;
+                HeadDomainNumber::<T>::set(domain_id, next_number);
```

Comment:
> If the ER never arrives for any reason, then Consensus chain says its n+1 without any ER

It surely can, because a domain block must be driven from a consensus block that contains the domain's bundle. This is what the consensus block can knows for sure.

Reply:
You are combining the state of Bundle submission and ER submission. If I think about primitives of the block chain itself, a head is only updated the execution of block at head is done. This is not the case here.

Clients may also fail at this point thinking Head is executed and for any proof verification and ends up failing instead. That is not a correct design IMO

Reply:
> You are combining the state of Bundle submission and ER submission

Submission of bundle and ER is always combined in both v1 and v2.

> If I think about primitives of the block chain itself, a head is only updated the execution of block at head is done.

The `HeadDomainNumber` is stored in the consensus chain, suppose there is a consensus block that contains extrinsic that updates a domain 's`HeadDomainNumber` from 1 to 2. Operators who have processed this consensus block (drive & import a domain block) its best block number update from 1 to 2, which match your above statement.

 > Clients may also fail at this point thinking Head is executed and for any proof verification and ends up failing instead. That is not a correct design IMO

Confused here, essentially `HeadDomainNumber` is just a storage item in the consensus chain, it won't change something like `client.info().best_number`

Reply:
> The HeadDomainNumber is stored in the consensus chain, suppose there is a consensus block that contains extrinsic that updates a domain 'sHeadDomainNumber from 1 to 2. Operators who have processed this consensus block (drive & import a domain block) its best block number update from 1 to 2, which match your above statement.

How do some one definitely prove that unless there is an execution receipt to corroborate that statement ?

Reply:
"Drive a domain block if the consensus block contains bundles" is not unknown for the consensus node and is not something we need to prove.

Reply:
Okay, I see the controversy here.

> just submitting domain bundle is not the correct step to consider a domain chain has progressed unless there is ER associated with it.

In my understanding, the domain chain is driven by the consensus chain so whenever it sees a consensus block that contains bundles, it drives and imports a domain block into the domain chain, so the chain progress at this moment. 

Due to the fact of the decouple execution architecture, the ER of the domain block is submitted to the consensus chain and tracked in the block tree, after the challenge period is passed the ER (namely the execution result of the domain block) is confirmed.

The `HeadReceiptNumber` is what we use to drive the challenge period moving forward thus it is apply to the whole block tree but not to any specific fork in the tree.

NOTE that the domain block **happen before** the ER, and the fork in the block tree is about the discrepancy of the execution result and is not the same as the chain fork. The term of `DomainBlock` and tree could be the source of confusion.